lib/ssl-test.rb

```diff
@@ -56,4 +64,98 @@ def self.matching_domains domains, hostname
     domains.map {|s| Regexp.new("\A#{Regexp.escape(s).gsub('\*', '[^.]+')}\z") }
       .select {|domain| domain.match?(hostname) }
   end
+
+  # https://docs.ruby-lang.org/en/2.2.0/OpenSSL/OCSP.html
+  # https://stackoverflow.com/questions/16244084/how-to-programmatically-check-if-a-certificate-has-been-revoked#answer-16257470
+  def self.test_ocsp_revocation cert, chain, open_timeout: 5, read_timeout: 5, redirection_limit: 5
+    issuer = chain.last
+
+    digest = OpenSSL::Digest::SHA1.new
+    certificate_id = OpenSSL::OCSP::CertificateId.new(cert, issuer, digest)
+
+    request = OpenSSL::OCSP::Request.new
+    request.add_certid certificate_id
+    request.add_nonce
+
+    authority_info_access = cert.extensions.find do |extension|
+      extension.oid == "authorityInfoAccess"
+    end
+
+    descriptions = authority_info_access.value.split("\n")
+    ocsp = descriptions.find do |description|
+      description.start_with?("OCSP")
+    end
+
+    ocsp_uri = URI(ocsp[/URI:(.*)/, 1])
+    http_response = follow_ocsp_redirects(ocsp_uri, request.to_der, open_timeout: open_timeout, read_timeout: read_timeout, redirection_limit: redirection_limit)
+    return OCSP_SOFT_FAIL_RETURN unless http_response
+
+    response = OpenSSL::OCSP::Response.new http_response.body
+    # https://ruby-doc.org/stdlib-2.6.3/libdoc/openssl/rdoc/OpenSSL/OCSP.html#constants-list
+    return OCSP_SOFT_FAIL_RETURN unless response.status == OpenSSL::OCSP::RESPONSE_STATUS_SUCCESSFUL
+    basic_response = response.basic
+
+    # Check the response signature
+    store = OpenSSL::X509::Store.new
+    store.set_default_paths
+    # https://ruby-doc.org/stdlib-2.4.0/libdoc/openssl/rdoc/OpenSSL/OCSP/BasicResponse.html#method-i-verify
+    return OCSP_SOFT_FAIL_RETURN unless basic_response.verify(chain, store)
```

Comment:
After fiddling with [OpenSSL::OCSP::BasicResponse#verify](https://ruby-doc.org/stdlib-2.4.0/libdoc/openssl/rdoc/OpenSSL/OCSP/BasicResponse.html#method-i-verify), I think we need to pass the full certificate chain as a first argument to validate the request, but TBH I'm not completely sure. WDYT?

Reply:
In the example on https://docs.ruby-lang.org/en/2.2.0/OpenSSL/OCSP.html they pass empty array, I guess you tried and it didn't work? As far as I understand here we're validating the OCSP response itself with the system store so I'm not sure why we would need the cert chain of the monitored website. But if it works like that and not without, I'm not sure I want to find out more xD

Reply:
> I guess you tried and it didn't work?

Yes. With an empty array, it fails, with the certificate chain, it works.

Let's leave it at that then ^^